File: core/src/main/java/io/squashql/query/join/QueryJoin.java
package io.squashql.query.join;

import io.squashql.query.dto.CriteriaDto;
import io.squashql.query.dto.JoinType;
import io.squashql.query.dto.QueryDto;
import io.squashql.query.dto.TableDto;
import lombok.NoArgsConstructor;

import java.util.ArrayList;
import java.util.List;

@NoArgsConstructor // For Jackson
public class QueryJoin {

  public List<QueryDto> queries = new ArrayList<>();
  public TableDto tableDto;
  public int current = 0;

  private QueryJoin(QueryDto q1) {
    this.queries.add(q1);
    this.tableDto = new TableDto(String.format("__cte%d__", this.current++));
  }

  public static QueryJoin from(QueryDto q) {
    return new QueryJoin(q);
  }

  public QueryJoin join(QueryDto q, JoinType joinType, CriteriaDto criteriaDto) {
    this.queries.add(q);
    this.tableDto.join(new TableDto(String.format("__cte%d__", this.current++)), joinType, criteriaDto);
    return this;
  }

  public QueryJoin join(QueryDto q, JoinType joinType) {
    this.queries.add(q);
    this.tableDto.join(new TableDto(String.format("__cte%d__", this.current++)), joinType, null);
    return this;
  }
}

File: js/typescript-library/src/queryJoin.ts
import {JoinType, Query, Table} from "./query"
import {serializeMap} from "./util"
import {Field} from "./field"
import {Order} from "./order"
import Criteria from "./criteria"

export class QueryJoin {

  minify?: boolean
  private readonly _table: Table
  private readonly _queries: Array<Query>
  private current: number = 0
  private _where: Criteria | null = null
  private _orders: Map<Field, Order>
  private _limit: number = -1

  constructor(query: Query) {
    this._queries = []
    this._queries.push(query)
    this._table = Table.from(`__cte${this.current++}__`)
    this._orders = new Map()
  }

  join(query: Query, joinType: JoinType, criteria?: Criteria): QueryJoin {
    this._queries.push(query)
    this._table.join(Table.from(`__cte${this.current++}__`), joinType, criteria)
    return this
  }

  where(where: Criteria) {
    this._where = where
    return this
  }

  orderBy(orders: Map<Field, Order>): QueryJoin {
    this._orders = orders
    return this
  }

  limit(limit: number) {
    this._limit = limit
    return this
  }

  toJSON() {
    return {
      "table": this._table,
      "queries": this._queries,
      "minify": this.minify,
      "where": this._where,
      "orders": Object.fromEntries(serializeMap(this._orders)),
      "limit": this._limit,
    }
  }
}
